src/parse.rs

```diff
@@ -0,0 +1,159 @@
+use nom::character::complete::{one_of, char, multispace1, u8, u64};
+use nom::bytes::complete::tag;
+use nom::branch::alt;
+use nom::combinator::{recognize, opt};
+use nom::sequence::{tuple, terminated, preceded, pair};
+use nom::IResult;
+use nom::multi::{count, many1};
+
+#[derive(PartialEq, Debug)]
+#[allow(dead_code)]
+
+pub enum FenError<'a> {
+    ParseErr(nom::Err<nom::error::Error<&'a str>>),
+    InvalidRow(Vec<char>),
+    InvalidActivePlayer(char),
+    InvalidRank(u8),
+    InvalidFile(u8)
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct ParsedGameState {
+    pub piece_position: [[char;8];8],
+    pub active_player: ActivePlayer,
+    pub castling_rights: CastlingRights,
+    pub en_passant_target: Option<(Rank, File)>,
+    pub half_turn_clock: u8,
+    pub full_turn_clock: u64
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[allow(dead_code)]
+pub enum ActivePlayer {
+    Black,
+    White
+}
+
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct CastlingRights {
+    pub black_kingside: bool,
+    pub black_queenside: bool,
+    pub white_kingside: bool,
+    pub white_queenside: bool
+}
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[repr(u8)]
+#[allow(dead_code)]
+pub enum Rank {
+    A = 0,
+    B = 1,
+    C = 2,
+    D = 3,
+    E = 4,
+    F = 5,
+    G = 6,
+    H = 7
+}
+
+impl TryFrom<u8> for Rank {
+    type Error = u8;
+    fn try_from(rank: u8) -> Result<Rank, Self::Error>{
+        match rank {
+            0..=7 => Ok(unsafe{std::mem::transmute(rank)}),
+            x => Err(x)
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[repr(u8)]
+#[allow(dead_code)]
+pub enum File {
+    One   = 0,
+    Two   = 1,
+    Three = 2,
+    Four  = 3,
+    Five  = 4,
+    Six   = 5,
+    Seven = 6,
+    Eight = 7
+}
+
+impl TryFrom<u8> for File {
+    type Error = u8;
+    fn try_from(rank: u8) -> Result<File, Self::Error>{
+        match rank {
+            0..=7 => Ok(unsafe{std::mem::transmute(rank)}),
+            x => Err(x)
+        }
+    }
+}
+
+pub fn fen_to_game(input: &str) -> Result<ParsedGameState, FenError>{
```

Comment:
We might want to break this out into several smaller functions, certainly at least one for piece positioning and perhaps another one for en passant.